### DISCUSSION TITLE
Support for junegunn/vim-plug to install

### DISCUSSION THREAD

**Discussion by mjj29** [CopilotC-Nvim/CopilotChat.nvim#46] [Q&A] [answered]
Hi there, I'm using vim-plug to manage my plugins, and it's working great for github/copilot, but installation failed (just seemed to hang) when I tried using it for this repository. Is there any way you can support that plugin manager as well?

Thanks,
Matt

**Comment by jellydn** [comment]
Hi @mjj29 There is a user confirmed and it's working for Vim Plug https://discord.com/channels/1200633211236122665/1200633212041449606/1205188460277665822

Let's ping him if he could help you on this.

**Reply by mjj29** [reply]
Yes, would be good to hear from someone else who has got it working (I don't have access to that discord server)

**Reply by jellydn** [reply]
Here you are @mjj29 https://discord.gg/vy6hJsTWaZ

**Comment by treyhunner** [comment, marked as answer]
I asked Chat GPT to translate the installation instructions for lazy.nvim to vim-plugin instead.

This is what I ended up with, which worked for me:

```
" Place this between call plug#begin() and call plug#end()
Plug 'CopilotC-Nvim/CopilotChat.nvim'

" After call plug#end(), configure the plugin options with VimScript
let g:copilotchat_options = {
      \ 'show_help': 'yes',
      \ 'debug': v:false,
      \ 'disable_extra_info': 'no',
      \ 'proxy': '',
      \ }

" Key mappings
nnoremap <leader>cce <cmd>CopilotChatExplain<cr>
nnoremap <leader>cct <cmd>CopilotChatTests<cr>
xnoremap <leader>ccv :CopilotChatVisual<cr>
xnoremap <leader>ccx :CopilotChatInPlace<cr>
```

After adding the plugin with vim-plug, run `:PlugInstall` in Neovim to install the CopilotChat.nvim plugin.

Next, the manual step of running `:UpdateRemotePlugins` is still necessary with vim-plug to ensure that Neovim properly recognizes the plugin's commands and functionalities. After running this command, restart Neovim.

To verify that the plugin has been correctly installed and initialized, you can check the `~/.local/share/nvim/rplugin.vim` file to see if the commands have been registered as expected.

Finally, restart Neovim to ensure all changes take effect and the plugin is ready to use.

**Reply by treyhunner** [reply]
Ah nevermind... it turns out the commands don't actually work after this. I'm getting errors about variables not being set:

```
  File "/home/trey/.dotfiles/config/nvim/plugged/CopilotChat.nvim/rplugin/python3/CopilotChat/handlers
/chat_handler.py", line 44, in chat
    self.nvim.eval("g:copilot_chat_disable_separators") == "yes"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
pynvim.api.common.NvimError: Vim:E121: Undefined variable: g:copilot_chat_disable_separators
```

And after setting that variable:

```
  File "/home/trey/.dotfiles/config/nvim/plugged/CopilotChat.nvim/rplugin/python3/CopilotChat/handlers
/chat_handler.py", line 46, in chat
    self.proxy = self.nvim.eval("g:copilot_chat_proxy")
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
pynvim.api.common.NvimError: Vim:E121: Undefined variable: g:copilot_chat_proxy
```

I don't think that options dictionary is actually correct.

**Reply by treyhunner** [reply]
After some Chat GPT-powered traceback debugging, this config seems to work!

```
" Place this between call plug#begin() and call plug#end()
Plug 'CopilotC-Nvim/CopilotChat.nvim'

let g:copilot_chat_disable_separators = "no"
let g:copilot_chat_proxy = ""
lua _COPILOT_CHAT_GLOBAL_CONFIG = _COPILOT_CHAT_GLOBAL_CONFIG or {}

" Key mappings
nnoremap <leader>cce <cmd>CopilotChatExplain<cr>
nnoremap <leader>cct <cmd>CopilotChatTests<cr>
xnoremap <leader>ccv :CopilotChatVisual<cr>
xnoremap <leader>ccx :CopilotChatInPlace<cr>
```

**Reply by mjj29** [reply]
OK, after a PlugClean then it got further and PlugInstall completed, and the config doesn't error, but I still can't actually use it, I get:

    E492: Not an editor command: CopilotChatInPlace

**Reply by treyhunner** [reply]
@mjj29 did you do `UpdateRemotePlugins`? I skipped that step at first and it didn't work until I ran that.

**Reply by jellydn** [reply]
Hi @treyhunner Could you share your setup on https://github.com/CopilotC-Nvim/CopilotChat.nvim/discussions/27?

**Reply by mjj29** [reply]
I hadn't, but that gives me this:

    function remote#host#UpdateRemotePlugins[6]..<SNR>30_RegistrationCommands[15]..remote#host#Require[10]..provider#pythonx#Require[12]..provider#Poll, line 7                                 
    Vim(if):Error invoking 'poll' on channel 3:^@ch 3 was closed by the client                                                                                                                  
    Traceback (most recent call last):                                                                                                                                                            
        File "<string>", line 1, in <module>                                                                                                                                                      
    ModuleNotFoundError: No module named 'neovim'                                                                                                                                               
    function remote#host#UpdateRemotePlugins[6]..<SNR>30_RegistrationCommands[15]..remote#host#Require[10]..provider#pythonx#Require[12]..provider#Poll, line 17                                
    Failed to load python3 host. You can try to see what happened by starting nvim with $NVIM_PYTHON_LOG_FILE set and opening the generated log file. Also, the host stderr is available in messages.                                                                                                                                                                                       
    remote/host: generated rplugin manifest: /users/ukcam/matj/.local/share/nvim/rplugin.vim

**Reply by jellydn** [reply]
> Failed to load python3 host

You should enable python3 on your Neonvim. You could test by run `py3 print("Hello")`.

**Reply by mjj29** [reply]
Hmm, what do I need to do for that?

    Package python3-neovim-0.4.3-1.el8.noarch is already installed.

**Reply by jellydn** [reply]
Here is my test for verify that's your Neonvim has enabled `python3` provider.

[![Image from Gyazo](https://i.gyazo.com/0de64d5861106929d9aa7b02099bb1a1.gif)](https://gyazo.com/0de64d5861106929d9aa7b02099bb1a1)

**Reply by mjj29** [reply]
OK, I got further by installing the following things with pip:

    pip3 install pynvim
    pip3 install python-dotenv
    pip3 install requests
    pip3 install prompt_toolkit

which let UpdateRemotePlugins run, but now I get:

    pynvim.api.common.NvimError: Vim:E121: Undefined variable: g:copilot_chat_user_prompts

**Reply by jellydn** [reply]
Could you share your setup here? @mjj29

**Reply by mjj29** [reply]
vimrc:

    call plug#begin('~/.vim/plugged')  
    ...
    Plug 'CopilotC-Nvim/CopilotChat.nvim'
    call plug#end()
    let g:copilot_chat_disable_separators = "no"
    let g:copilot_chat_proxy = ""
    lua _COPILOT_CHAT_GLOBAL_CONFIG = _COPILOT_CHAT_GLOBAL_CONFIG or {}
    nnoremap <leader>cce <cmd>CopilotChatExplain<cr>
    nnoremap <leader>cct <cmd>CopilotChatTests<cr>
    xnoremap <leader>ccv :CopilotChatVisual<cr>
    xnoremap <leader>ccx :CopilotChatInPlace<cr>
    ...

Contents of ~/.local/lib/python3.9/site-packages:

    $ ls ~/.local/lib/python3.9/site-packages/
    certifi/                            msgpack/                            requests-2.31.0.dist-info/
    certifi-2024.2.2.dist-info/         msgpack-1.0.7.dist-info/            urllib3/
    charset_normalizer/                 neovim/                             urllib3-2.2.0.dist-info/
    charset_normalizer-3.3.2.dist-info/ prompt_toolkit/                     wcwidth/
    dotenv/                             prompt_toolkit-3.0.43.dist-info/    wcwidth-0.2.13.dist-info/
    greenlet/                           pynvim/                             wheel/
    greenlet-3.0.3.dist-info/           pynvim-0.5.0.dist-info/             wheel-0.42.0.dist-info/
    idna/                               python_dotenv-1.0.1.dist-info/
    idna-3.6.dist-info/                 requests/

I've run :PlugInstall and :UpdateRemotePlugins, both of which completed successfully now, but when I run :CopilotChatInPlace I get:

    error caught in async handler '/users/ukcam/matj/.vim/plugged/CopilotChat.nvim/rplugin/python3/CopilotChat:command:CopilotChatInPlace [[], [1, 1]]'
    Traceback (most recent call last):
      File "/users/ukcam/matj/.vim/plugged/CopilotChat.nvim/rplugin/python3/CopilotChat/copilot_plugin.py", line 64, in inplace_cmd
        self.init_inplace_chat_handler()
      File "/users/ukcam/matj/.vim/plugged/CopilotChat.nvim/rplugin/python3/CopilotChat/copilot_plugin.py", line 49, in init_inplace_chat_handler
        self.inplace_chat_handler = InPlaceChatHandler(self.nvim)
      File "/users/ukcam/matj/.vim/plugged/CopilotChat.nvim/rplugin/python3/CopilotChat/handlers/inplace_chat_handler.py", line 25, in __init__
        self.user_prompts = self.nvim.eval("g:copilot_chat_user_prompts")
      File "/users/ukcam/matj/.local/lib/python3.9/site-packages/pynvim/api/nvim.py", line 319, in eval
        return self.request('nvim_eval', string, **kwargs)
      File "/users/ukcam/matj/.local/lib/python3.9/site-packages/pynvim/api/nvim.py", line 199, in request
        res = self._session.request(name, *args, **kwargs)
      File "/users/ukcam/matj/.local/lib/python3.9/site-packages/pynvim/msgpack_rpc/session.py", line 139, in request
        raise self.error_wrapper(err)
    pynvim.api.common.NvimError: Vim:E121: Undefined variable: g:copilot_chat_user_prompts

os version Red Hat Enterprise Linux release 8.9 (Ootpa)
nvim version NVIM v0.8.0
python version Python 3.9.18
node version v18.18.2

**Reply by jellydn** [reply]
Could you add those to Plug? @mjj29 

```lua
Plug 'CopilotC-Nvim/CopilotChat.nvim'
call plug#end()

let g:copilot_chat_disable_separators = "no"
let g:copilot_chat_proxy = ""
lua _COPILOT_CHAT_GLOBAL_CONFIG = _COPILOT_CHAT_GLOBAL_CONFIG or {}
let g:copilot_chat_user_prompts = {
  Explain = 'Explain how it works.',
  Tests = 'Briefly explain how selected code works then generate unit tests.',
}
let g:copilot_chat_show_help = 'yes'
let copilot_chat_disable_separators = 'no'
let copilot_chat_hide_system_prompt = 'yes' 

nnoremap <leader>cce <cmd>CopilotChatExplain<cr>
nnoremap <leader>cct <cmd>CopilotChatTests<cr>
xnoremap <leader>ccv :CopilotChatVisual<cr>
xnoremap <leader>ccx :CopilotChatInPlace<cr>
```

**Reply by pietro1704** [reply]
Is this on `~/.config/nvim/vim-plugs/plugins.vim`? For me, this code after `call plug#end` doesn't get recognized by `nvim`. I'm using `vim-plug` on `Nvim`.

**Comment by nekowasabi** [comment]
My neovim's setting with vim-plug.
I have CopilotChat working with this configuration.

```
lua << EOF
local copilot_chat = require("CopilotChat")
copilot_chat.setup({
  debug = true,
  show_help = "yes",
  prompts = {
    Explain = "Explain how it works by Japanese language.",
    Review = "Review the following code and provide concise suggestions.",
    Tests = "Briefly explain how the selected code works, then generate unit tests.",
    Refactor = "Refactor the code to improve clarity and readability.",
		Aaa = "hello",
  },
  build = function()
    vim.notify("Please update the remote plugins by running ':UpdateRemotePlugins', then restart Neovim.")
  end,
  event = "VeryLazy",
})

EOF

" CopilotChat
vnoremap <silent> <leader>cv :CopilotChatVisual 
vnoremap <silent> <leader>cx :CopilotChatInPlace<CR>
vnoremap <silent> <leader>ce :CopilotChatExplain<CR>
```
